monkey/infection_monkey/exploit/tools/brute_force_exploiter.py

```diff
@@ -0,0 +1,209 @@
+
+    def _exploit(self, exploit_client: IRemoteAccessClient, host: TargetHost, interrupt: Event):
+        credential_combinations = self._get_credentials()
+
+        for brute_force_credentials in interruptible_iter(credential_combinations, interrupt):
+            tags: MutableSet[str] = set()
+            try:
+                exploit_client.login(brute_force_credentials, tags)
+                self._publish_exploitation_event(host, success=True, tags=self._tags.union(tags))
+                return
+            except RemoteAuthenticationError:
+                self._publish_exploitation_event(host, success=False, tags=self._tags.union(tags))
+                continue
+
+        raise Exception("Failed to login with the given credentials")
+
+    def _propagate(
+        self,
+        exploit_client: IRemoteAccessClient,
+        host: TargetHost,
+        interrupt: Event,
+    ):
+        target_host_os = exploit_client.get_os()
+        agent_binary = self._agent_binary_repository.get_agent_binary(target_host_os)
+        file_path = self._copy_file(
+            agent_binary, self._destination_path, exploit_client, host, interrupt
+        )
+
+        command = self._build_command(target_host_os, file_path, self._destination_path)
+        tags: MutableSet[str] = set()
+        try:
+            exploit_client.execute_detached(command, tags)
+            self._publish_propagation_event(host, success=True, tags=self._tags.union(tags))
+        except RemoteCommandExecutionError:
+            self._publish_propagation_event(host, success=False, tags=self._tags.union(tags))
+            raise
+
+    def _copy_file(
+        self,
+        file: BytesIO,
+        destination: PurePath,
+        exploit_client: IRemoteAccessClient,
+        host: TargetHost,
+        interrupt: Event,
+    ) -> PurePath:
+        file_data = file.getvalue()
+        tags: MutableSet[str] = set()
+        try:
+            exploit_client.copy_file(file_data, destination, tags)
+            return destination
+        except RemoteFileCopyError as err:
+            self._publish_exploitation_event(
+                host,
+                success=False,
+                tags=self._tags.union(tags),
+                error_message=str(err),
+            )
+
+        other_destinations = exploit_client.get_writable_paths()
+        for other_destination in interruptible_iter(other_destinations, interrupt):
+            tags = set()
+            try:
+                exploit_client.copy_file(file_data, other_destination, tags)
+                return other_destination
+            except RemoteFileCopyError as err:
+                self._publish_exploitation_event(
+                    host,
+                    success=False,
+                    tags=self._tags.union(tags),
+                    error_message=str(err),
+                )
+
+        raise Exception("Failed to copy file")
```

Comment:
Can we reduce duplication by just surfacing all errors up to the calling function? We can have the calling function pass the tags output parameter down to this one.

Reply:
I didn't see a unit test that verifies only copy tags get added if copy fails.